### PR TITLE
Expose fairness RSS structure metrics

### PR DESCRIPTION
**Pull request by psaab**
## Summary
- add production Prometheus gauges derived from the existing per-CoS active-flow snapshot: `xpf_fairness_cstruct`, `xpf_fairness_active_flows`, `xpf_fairness_active_workers`, and `xpf_fairness_max_worker_flow_share`
- add `xpf_fairness_cos_active_flow_counts_truncated` so operators can tell when the source snapshot was capped
- compute Cstruct from the `{a_i}` distribution with a weighted streaming variance, without allocating per-flow share vectors and without adding packet-path state or global atomics
- update fairness docs and Path 4 state to distinguish shipped RSS-structure gauges from still-open rolling observed-CoV/starvation windows
- align future rolling fairness metric labels on `{ifindex, queue_id}` and document the truncation gauge in the Path 4 state

## Validation
- `go test ./pkg/api`
- `go test ./pkg/api -run 'TestEmit(FairnessRSSGauges|CoSActiveFlowCount)|TestFairnessRSSAggregateCstruct'`
- `git diff --check`

Follow-up to #1247. This does not add opt-in runtime alert config yet; it makes the structural RSS state observable in production first.

### CODE REVIEW

**Comment by psaab** [comment]
## Claude round-1 review on `8f00f5ce`

**Verdict: MERGE-READY**

This PR exposes 5 production Prometheus gauges derived from the existing per-CoS active-flow snapshot, with a closed-form streaming Cstruct calculation that avoids per-flow allocations. Math is correct and verified against the Rust reference; tests are sound; diff hygiene is clean.

### Math verification — closed-form vs Rust reference

The Rust reference `compute_cstruct` at `userspace-dp/src/fairness.rs:24` builds the per-flow share multiset `{1/a_i repeated a_i times}` and computes population CoV directly. The Go code uses a closed-form streaming equivalent:

```go
// Go (closed-form, no per-flow vec):
mean         = activeWorkers / totalFlows
secondMoment = sumInverseFlows / totalFlows         // = E[shares²]
variance     = secondMoment - mean²                 // = Var(shares)
cstruct      = sqrt(variance) / mean
```

**Algebraic equivalence proof:**
- `sum(shares) = sum_i a_i × (1/a_i) = activeWorkers` → `mean = activeWorkers / totalFlows` ✓
- `sum(shares²) = sum_i a_i × (1/a_i)² = sum_i (1/a_i) = sumInverseFlows`
- `E[shares²] = sumInverseFlows / totalFlows` ✓
- `Var(X) = E[X²] − E[X]²` ✓

**Worked example — test case Q4 `[1, 3, 0]`** (asserts `0.577350269`):
- activeWorkers=2, totalFlows=4, sumInverseFlows = 1 + 1/3 = 4/3
- mean = 2/4 = 0.5
- secondMoment = (4/3)/4 = 1/3
- variance = 1/3 − 1/4 = 1/12
- cstruct = √(1/12) / 0.5 = **1/√3 ≈ 0.57735** ✓

**Q5 uniform `[2, 2]`** (asserts 0):
- activeWorkers=2, totalFlows=4, sumInverseFlows = 1.0
- mean = 0.5, secondMoment = 0.25, variance = 0 ✓

### Hostile-check findings

**Numerical stability** (the catastrophic-cancellation concern)
The closed-form `E[X²] − E[X]²` is classically unstable for low-variance inputs. Worst-case worked example — `[9, 11]`:
- secondMoment = 0.010101, mean² = 0.01, variance = 0.000101
- Loses ~2 digits of precision to cancellation, but with double precision (15-17 digits), 13 reliable digits remain.

For monitoring purposes where Cstruct feeds a contract threshold comparison (e.g., `observed_CoV ≤ Cstruct + 0.05`), this precision is more than adequate. The `if variance <= 0 { return 0 }` guard correctly handles the fp-rounding-produces-tiny-negative case for perfectly-uniform distributions like `[3,3,3]` where secondMoment and mean² should exactly cancel. ✓

**Hot-path / allocation** — per-scrape map allocation (`byQueue` + `keys` slice). With ~5-10 ifindex × 8 CoS queues, bounded at ~80 entries. Prometheus scrape cadence is 15s — trivially acceptable. ✓

**Label cardinality** — `(ifindex, queue_id)` is bounded. ✓

**Counter overflow** — `totalActiveFlows uint64`, `activeWorkers uint64`, `maxWorkerFlows uint32`, `sumInverseFlows float64`. All bounded for realistic configs. ✓

**Sort order** — keys sorted by ifindex then queue_id. Matches existing `emitCoSActiveFlowCount` pattern; emission stable across scrapes. ✓

**Snapshot consistency** — `CoSActiveFlowCounts` and `CoSActiveFlowCountsTruncated` come from a single `dp.GetUserspaceStatus()` call. Consistent with the rest of the collector's snapshot model. ✓

**Test coverage** — tests build `xpfCollector` with `NewDesc` directly rather than via `newCollector()`. Matches the existing test pattern for `TestEmitCoSActiveFlowCount_LabelsAndValue`. Coverage is the math, not the wiring — wiring bugs in `newCollector` would surface elsewhere.

### Minor observations (not blocking)

**A. Dead-code check `if mean == 0`** (line ~580). Already short-circuited by the earlier `if totalActiveFlows == 0 || activeWorkers == 0 { return 0 }`. Stylistic redundancy.

**B. `fairnessCoSCountsTruncated` is a snapshot-level singleton boolean.** If the truncation source is per-binding rather than per-queue (which it is per #1248's `BindingLiveState.flow_worker_map.truncated`), this single gauge collapses the per-binding truncation flags into one OR. Documented as snapshot-level, so acceptable, but operators may want per-(ifindex, queue) granularity in a follow-up.

**C. No cross-test against the Rust reference.** The Go test asserts numeric values directly. Ideally there'd be a property test that generates random distributions and asserts Go cstruct == Rust cstruct ± epsilon. Future-proofing concern; not a regression.

### Recommendation

Ready to merge. Awaiting Codex (`task-mp1m6gha-me6ltn`) and Gemini Pro 3 (`task-mp1m6yuv-ez84tm`) for full triple-review confirmation. Copilot has not reviewed this commit yet; will trigger after Codex/Gemini land.

**Comment by psaab** [comment]
## Round-1 triple-review consolidated synthesis on `8f00f5ce`

| Reviewer | Verdict |
|---|---|
| Claude | MERGE-READY |
| Codex | **MERGE-NEEDS-MINOR** (numerical-stability concern, edge-case test gap) |
| Gemini Pro 3 | ✅ MERGE-READY (no hallucinations this round — verified line-by-line) |
| Copilot | 2 inline (both valid, both doc-only) |

### Math correctness — all three reviewers verified equivalence

Cross-verified the closed-form Go implementation against the Rust reference `compute_cstruct` at `userspace-dp/src/fairness.rs:24`. All reviewers ran the math by hand for `[1,3,0]` and got `1/√3 ≈ 0.5774`, matching the test assertion. Codex also pinned `[2,2]=0`, `[1]=0`, `[5]=0`, `[3,3,3]=0`, `[1,99]=4.92468529477`. ✓

### Codex finding — numerical stability (real math, operationally unreachable)

Codex flagged `metrics.go:578-583` as catastrophic-cancellation:
> Example representable inputs `[1_000_000_000, 1_000_000_001]` have true `cstruct ~= 5e-10`, but the closed-form subtraction rounds to zero. `[uint32_max-1, uint32_max]` also emits zero for a nonzero value.

**Cross-check:** Real math, verified. The closed-form `variance = secondMoment - mean²` is the classic catastrophic-cancellation pattern.

**Operational reach:** In this codebase, `FLOW_WORKER_MAP_MAX_PER_BINDING = 256`. With ~16 workers × ~256 flows each, max active flows per worker stays in the low hundreds. Realistic mean ≈ 1e-3 to 1e-4, secondMoment ≈ 1e-6 to 1e-8 — well above the fp cancellation floor (~1e-15 relative). Codex's billion-scale example is operationally unreachable.

**Suggested fix (Codex)**: weighted streaming variance — `sum_i active_i × (1/active_i - mean)² / totalFlows`. One-line alternative that avoids cancellation entirely.

My read: not blocking for monitoring purposes (alert threshold is `Cstruct + 0.05`, vastly larger than the precision floor), but worth folding into the same PR as a code-quality improvement since the fix is trivial. **Reasonable to defer to follow-up** if author prefers.

### Codex finding — edge-case test coverage gap

Tests cover Q4 `[1,3,0]` and Q5 `[2,2]` but not the canonical edge table Codex suggested in the prompt: `[1]`, `[5]`, `[3,3,3]`, `[1,99]`, near-uniform stability case. Direct table-driven tests on `fairnessRSSAggregate.cstruct()` would lock the contract.

Minor; the existing tests verify the math correctly via end-to-end emission. Adding the edge table is defense against future regression.

### Copilot inline #1 — `docs/fairness-regimes.md:240,255,257` (VALID)

Planned (not-yet-implemented) metrics `xpf_fairness_saturated`, `xpf_fairness_observed_cov`, `xpf_fairness_starved_flows` still document `{queue=...}` label, but the gauges added in this PR use `{ifindex, queue_id}`. Future-implementation inconsistency.

**Cross-check verified at `docs/fairness-regimes.md`:240,255,257** — confirmed.

**Fix**: update the not-yet-implemented entries to `{ifindex, queue_id}` to align future implementations with the shipped contract.

### Copilot inline #2 — `docs/per-5-tuple/state.md:325-327` (VALID)

Lists 4 of 5 newly-exported gauges (cstruct, active_flows, active_workers, max_worker_flow_share) but omits `xpf_fairness_cos_active_flow_counts_truncated`, which is part of the same observability slice.

**Cross-check verified at `docs/per-5-tuple/state.md`:325-327** — confirmed; truncation gauge missing from the list.

**Fix**: add `xpf_fairness_cos_active_flow_counts_truncated` to the same enumeration so operators don't miss the truncation signal.

### Gemini Pro 3 — no hallucinations

Gemini verified all 10 checks line-by-line, including the algebraic equivalence of the closed-form vs Rust reference. Explicitly noted the catastrophic-cancellation guard `if variance <= 0 { return 0 }` is "mathematically sound (as the structural skew *is* 0)" for perfectly-uniform distributions, and that "precision loss is irrelevant for Prometheus alerting" near zero. No false claims. Clean round.

### Recommendation

**Strongly consider in this PR (low-effort, high-clarity)**:
1. Copilot #1 — align label set in `fairness-regimes.md:240,255,257` planned metrics
2. Copilot #2 — list truncation gauge in `per-5-tuple/state.md:325-327`

**Defer to follow-up (operationally unreachable)**:
3. Codex numerical-stability — closed-form cancellation at billion-scale flow counts; switch to weighted streaming variance when convenient
4. Codex edge-case test table — defense against future regression

Codex task: `task-mp1m6gha-me6ltn`. Gemini Pro 3 task: `task-mp1m6yuv-ez84tm`. Not merging.

File: docs/per-5-tuple/state.md [hunk 1]
```diff
@@ -320,6 +320,14 @@ checked at runtime, not just in test.
 violates that expectation. Runtime config/alerting can build on the
 same expectation language.
 
+The second production-observability slice exports per-CoS RSS structure
+gauges from the existing userspace status snapshot:
+`xpf_fairness_cstruct`, `xpf_fairness_active_flows`,
+`xpf_fairness_active_workers`, and
+`xpf_fairness_max_worker_flow_share`. These make structural skew visible
+in Prometheus without adding packet-path state. Opt-in config/alerting
+for expectation violations remains separate follow-up work.
```
[the remark this answers is not in the document]
Comment (psaab) [reply docs/per-5-tuple/state.md]: Fixed in c19e47a8. `docs/per-5-tuple/state.md` now lists `xpf_fairness_cos_active_flow_counts_truncated` with the other RSS-structure gauges and describes it as the capped-source-snapshot signal.

File: docs/fairness-regimes.md [hunk 2]
```diff
@@ -331,11 +348,9 @@ For production observability, xpf MUST export:
...
-  underlying signal and is exported separately if the harness
-  needs it for context.
-- **`xpf_fairness_cstruct{queue=...}`** gauge: the current
-  computed structural CoV ceiling.
+  the per-worker active-flow distribution `{a_i}` is the underlying
+  signal and is exported separately if the harness needs it for
+  context.
 - **`xpf_fairness_observed_cov{queue=...}`** gauge: rolling
   observed CoV for the queue.
 - **`xpf_fairness_starved_flows{queue=...}`** counter:
```
[the remark this answers is not in the document]
Comment (psaab) [reply docs/fairness-regimes.md]: Fixed in c19e47a8. The planned rolling-window metrics now use `{ifindex=..., queue_id=...}` in `docs/fairness-regimes.md`, matching the shipped per-CoS RSS-structure gauges and avoiding a future label-set fork.

**Comment by psaab** [comment]
Addressed round-1 review feedback in c19e47a8:

- Switched `fairnessRSSAggregate.cstruct()` from `E[X²] - E[X]²` to weighted streaming variance, avoiding the near-uniform catastrophic-cancellation case while still avoiding per-flow share-vector allocation.
- Added edge-case coverage for `[1]`, `[5]`, `[3,3,3]`, `[1,99]`, `[1_000_000_000,1_000_000_001]`, and `[uint32_max-1,uint32_max]`.
- Aligned planned rolling metric labels in `docs/fairness-regimes.md` to `{ifindex, queue_id}`.
- Added `xpf_fairness_cos_active_flow_counts_truncated` to the Path 4 state gauge list.

Validation: `go test ./pkg/api`, focused fairness metric tests, and `git diff --check`.

**Comment by psaab** [comment]
## Claude round-2 review on `c19e47a8` (force-pushed)

**Verdict: MERGE-READY**

Comprehensive follow-up that addresses every round-1 finding with a textbook-correct numerical fix.

### Round-1 → round-2 fix verification

| # | Round-1 finding | Round-2 fix | Status |
|---|---|---|---|
| 1 | **Codex MINOR** — closed-form `variance = E[X²] − E[X]²` collapses near-uniform billion-scale to 0 | Replaced with **Welford's weighted online algorithm** in `add()` | ✅ Textbook correct |
| 2 | **Codex MINOR** — no edge-case tests for `[1]`, `[5]`, `[3,3,3]`, `[1,99]`, near-uniform | New `TestFairnessRSSAggregateCstruct_EdgeCases` with 6 cases (incl. Codex's billion-scale + u32-max repros) | ✅ All pass |
| 3 | **Copilot** — `docs/fairness-regimes.md` planned metrics used `{queue=...}` label | Now `{ifindex=..., queue_id=...}` | ✅ |
| 4 | **Copilot** — `docs/per-5-tuple/state.md` missing truncation gauge | Now lists `xpf_fairness_cos_active_flow_counts_truncated` alongside the other 4 | ✅ |

### Welford weighted variance — verified correct

The new `add()` implements weighted Welford correctly:
```go
weight := float64(active)
value := 1.0 / weight
oldTotal := float64(a.totalActiveFlows)
newTotal := oldTotal + weight
delta := value - a.weightedMean
a.weightedMean += (weight / newTotal) * delta
a.weightedM2 += weight * delta * (value - a.weightedMean)  // mean post-update
```

Then `variance = weightedM2 / totalActiveFlows`, `cstruct = √variance / weightedMean`.

**Cross-verified with the analytical formula** for `[N, N+1]` distributions:
- shares multiset: N copies of `1/N`, N+1 copies of `1/(N+1)`
- After simplification: **`cstruct = 1 / (2·√(N·(N+1)))`**

For N=10⁹ (the billion-scale case Codex flagged): cstruct ≈ 1/(2·√(10¹⁸+10⁹)) = **5·10⁻¹⁰**. Asserted: `4.9999999975e-10`. ✓

For N=u32_max-1 (uint32-max case): cstruct ≈ 1/(2·u32_max) ≈ **1.16415e-10**. Asserted: `1.16415321868e-10`. ✓

The author got the analytical math right AND the asserted test values match an independent derivation. Real fix, not a rebranded closed-form.

### Test execution

Ran in worktree at `c19e47a8`:
```
=== RUN   TestEmitFairnessRSSGauges_DerivesStructuralCeiling      --- PASS
=== RUN   TestEmitFairnessRSSGauges_EmptyDistributionOnlyReportsTruncation --- PASS
=== RUN   TestFairnessRSSAggregateCstruct_EdgeCases               --- PASS (6 subtests)
```

All 8 tests pass. Q4 `[1,3,0]=0.5774` and Q5 `[2,2]=0` still produce identical values to the original closed-form, confirming Welford gives the same answer in the operationally-realistic small-flow regime.

### Diff hygiene

4 files only, +56/-16. No unrelated churn:
- `pkg/api/metrics.go` (+23/-12 — Welford rewrite, public signature unchanged)
- `pkg/api/metrics_test.go` (+35 — edge-case table)
- `docs/fairness-regimes.md` (+3/-3 — label set)
- `docs/per-5-tuple/state.md` (+5/-3 — truncation gauge in list)

### Recommendation

Ready to merge. Awaiting Codex (`task-mp1mmx81-i22rk1`) and Gemini Pro 3 (`task-mp1mndtn-1jo31h`) for full triple-review confirmation. Copilot has not yet reviewed `c19e47a8` (last review on `8f00f5ce`); will request re-review after Codex/Gemini land.

**Comment by psaab** [comment]
@copilot review

Round-2 commit `c19e47a8` addresses all four round-1 findings:

1. **Numerical-stability (Codex MINOR)**: replaced closed-form variance with Welford weighted online algorithm. Independently verified against the analytical formula `cstruct = 1/(2·√(N·(N+1)))` for [N, N+1] distributions. New edge-case tests cover billion-scale and u32-max inputs.

2. **Edge-case test coverage (Codex MINOR)**: TestFairnessRSSAggregateCstruct_EdgeCases pins [1], [5], [3,3,3], [1,99], [1e9, 1e9+1], [u32_max-1, u32_max].

3. **fairness-regimes.md label set (Copilot)**: planned metrics now use `{ifindex=..., queue_id=...}` instead of `{queue=...}`.

4. **per-5-tuple/state.md (Copilot)**: now lists `xpf_fairness_cos_active_flow_counts_truncated` alongside the other 4 gauges.

Three-of-three reviewer consensus: MERGE-READY (Claude + Codex + Gemini Pro 3).

**Comment by psaab** [comment]
## Round-2 triple-review consolidated synthesis on `c19e47a8`

| Reviewer | Verdict |
|---|---|
| Claude | ✅ MERGE-READY |
| Codex | ✅ **MERGE-READY** (no findings — verified Welford correctness independently) |
| Gemini Pro 3 | ✅ **MERGE-READY** (verified analytically — Taylor expansion + ULP analysis) |
| Copilot | Re-review requested |

### Three-of-three reviewer consensus: MERGE-READY

All four round-1 findings cleanly addressed:

| # | Round-1 finding | Round-2 fix | Verification |
|---|---|---|---|
| 1 | Codex MINOR — closed-form variance cancels at billion-scale | Welford weighted online algorithm | All 3 reviewers independently verified math |
| 2 | Codex MINOR — no edge-case tests | 6-row table including Codex's exact repros | Tests pass; values match analytical formula |
| 3 | Copilot — `{queue=...}` label inconsistency in docs | Now `{ifindex=..., queue_id=...}` | Verified in `docs/fairness-regimes.md:341,354,356` |
| 4 | Copilot — missing truncation gauge in `state.md` | Now listed | Verified in `docs/per-5-tuple/state.md:327-329` |

### Independent math verification (multiple reviewers, same conclusion)

For `[N, N+1]` distribution, analytical Cstruct simplifies to `1 / (2·√(N·(N+1)))`:

- **N = 10⁹** → `4.9999999975 × 10⁻¹⁰` (Taylor: `0.5e-9 - 0.25e-18`)
  - Welford float64 returns `5.000000749806408 × 10⁻¹⁰`
  - Test asserts `4.9999999975e-10` within `1e-12` tolerance → **passes**
- **N = u32_max-1** → `1.164153218675924 × 10⁻¹⁰`
  - Welford float64 returns `1.164153217930535 × 10⁻¹⁰`
  - Test asserts `1.16415321868e-10` within `1e-12` tolerance → **passes**

### Notable reviewer quality this round

- **Gemini Pro 3 followed the verify-before-claiming guidance.** Verified each algebraic claim against the code at line numbers, performed Taylor expansion to derive the asserted Welford output, analyzed ULP behavior to justify the `variance <= 0` guard. **No hallucinations.**
- **Codex independently computed exact references** at 64-bit precision for all asserted Cstruct values and confirmed they fall within `1e-12` of Welford's float64 result.

### Recommendation

**Three-of-three reviewer consensus. Ready to merge** once Copilot confirms (re-review requested above). All deferred-to-follow-up items from round-1 are now addressed in this PR.

Codex task: `task-mp1mmx81-i22rk1`. Gemini Pro 3 task: `task-mp1mndtn-1jo31h`. Not merging — author's decision.
